### ISSUE TITLE
examples/hello (and others) creates empty window, but no rendering

### ISSUE THREAD

**Issue by mewmew** [lakal3/VGE#10]
I tried to get the examples running, and with the merged PRs they are now successfully compiling.

However, when I try to run an example it opens a window, but does not seem to draw anything into it. (**note:** the window just gets the contents of the background, so, I switched wallpaper to make the VGE window more visible.)

![2021-03-22-212127_1089x825_scrot](https://user-images.githubusercontent.com/1414531/112053295-76b4b480-8b54-11eb-90f9-8bd787327042.png)

I tried vkcube and it is working correctly.
![2021-03-22-212023_829x577_scrot](https://user-images.githubusercontent.com/1414531/112053277-71576a00-8b54-11eb-8f2a-2e07313b7de4.png)

Any idea how to go about troubleshooting this? I'm still very new to Vulkan and would love to dive in and learn more. That's why I felt to happy to discover your work on VGE!

Cheers,
Robin

**Comment by mewmew** [comment]
When debugging, it seems `GetNextFrame` in e.g. the cube example is returning `-1`. Exactly why I'm still not sure yet.

**Comment by lakal3** [comment]
You could do several tests to narrow down actual problem.

**1. Run basic tests**

Set VGE_TEST_DIR environment variable and run go test in vge/vk module. You should get no errors and vk.png image to {VGE_TEST_DIR}. 
Image should show a triangle with a test pattern. If this won't work, please post some more details of error/problem.

**2. Run cube example**

Build cube example and run it. It will write display driver name on console. If you have multiple display drivers that at least most 
laptops have check try to change display driver with -dev switch. At least on my old Linux, only dedicated GPU was able to create a swap chain that is needed to display rendered images on screen.
If switching devices don't work, try step 3.

**3. Run webview example**

Build webview example. It will not use your screen directly to display image. Instead it have small web server that will generate image in background using Vulkan API and then uses web browser to display rendered images. This demo should work even if we can't create a swap chain. Webview example has also -dev switch that can be use to change Vulkan device if your system has multiple Vulkan devices.

**Comment by mewmew** [comment]
Great! Thanks for helping to trouble-shoot this @lakal3. Many of the tests seem to be working:

![vk](https://user-images.githubusercontent.com/1414531/112207317-42ee9300-8c17-11eb-9934-f47da03870b6.png)

It seems some test cases are using old locations for assets:

```
[u@x1 vge]$ go test -count=1 ./...
# github.com/lakal3/vge/vge/dldyn
dldyn/dl.go:1:1: +build comment must appear before package clause and be followed by a blank line
ok  	github.com/lakal3/vge/vge/materials/decal	0.125s
ok  	github.com/lakal3/vge/vge/materials/noise	0.021s
ok  	github.com/lakal3/vge/vge/vanimation	0.063s
ok  	github.com/lakal3/vge/vge/vapp	0.021s
# github.com/lakal3/vge/vge/vui/mintheme [github.com/lakal3/vge/vge/vui/mintheme.test]
vui/mintheme/theme_test.go:32:23: gs.GetImage undefined (type *vglyph.GlyphSet has no field or method GetImage)
ok  	github.com/lakal3/vge/vge/vglyph	0.243s
ok  	github.com/lakal3/vge/vge/vk	0.113s

--- FAIL: TestNewUIView (0.01s)
    testapp.go:23: API call failed:  open /home/u/Desktop/VGE/assets/glyphs/basicui/btn.png: no such file or directory
FAIL
FAIL	github.com/lakal3/vge/vge/vui	0.014s
FAIL	github.com/lakal3/vge/vge/vui/mintheme [build failed]
```

**Comment by mewmew** [comment]
The webview example works. There is only one device available on my laptop it seems.

```
$ env GODEBUG=cgocheck=0 LD_LIBRARY_PATH=. ./webview -dev 1
2021/03/23 20:43:47 No device 1
```

```
$ env GODEBUG=cgocheck=0 LD_LIBRARY_PATH=. ./webview -dev 0

Using device  Intel(R) HD Graphics 620 (KBL GT2)
Running on localhost:3050
```

![screenshot_2021-03-23_20:44:25](https://user-images.githubusercontent.com/1414531/112208202-6108c300-8c18-11eb-9a89-d0c0433b8a23.png)

**Comment by mewmew** [comment]
I tried the cube example (using the only device available: `-dev 0` *Intel(R) HD Graphics 620 (KBL GT2)*), but it has the same problem as the `hello` example above. The window opens, but no rendering occur. The call to `GetNextFrame` returns `-1`.

Window events are working perfectly though. Any idea how to continue trouble-shooting?

![2021-03-23-205446_1221x1013_scrot](https://user-images.githubusercontent.com/1414531/112210685-4b48cd00-8c1b-11eb-93b1-3bf191963191.png)

**Comment by mewmew** [comment]
By the way, I tried building VGE on a Windows machine yesterday and it worked perfectly. It's incredible to see the 3D environments with lighting and model animations running spotless and so smoothly using Go! Figuratively, I take my hat of and bow to you @lakal3, you've really achieved something incredible with VGE and I think it is a solid foundation to keep developing on!

With excited admiration,
Robin

**Comment by mewmew** [comment]
Lauri, you wouldn't happen to have Discord? If so, I would be happy to get the chance to chat with you, should you feel like of course.

**Comment by lakal3** [comment]
Thanks for encouraging feedback :). 

Sorry, I don't have discord at the moment. However, I could try to install one after Easter holidays. 

Actually I started this project when I wanted to learn Vulkan API and 3D programming technologies. I initially used C++ but if was so frustrated with pure C++ that I decided to try hybrid approach with only minimal C++ code. This worked better than I ever expected. 
It is bit sad that Golang don't support similar DLL (so) loading in Linux that works in Windows. In most other areas Golang cross OS support is excellent.

**Comment by lakal3** [comment]
For the matter in hand. Could you try to debug PrepareSwapchain function in desktop.cpp file (around line 162). It will call createSwapchain function that should create swapchain images but for some reason we end up with zero images. I do have a machine with Intel chip but that laptop has Window and in Window this createSwapchain seems to work nicely.

I have taken some shortcuts in createSwapchain that might be biting here.

If you have time, you could check https://github.com/SaschaWillems/Vulkan_Test/blob/master/01_triangle/01_triangle.cpp, pure C++ demo and see if that works on you Linux machine. If that works, could you pinpoint what is done 01_triangle.cpp demo that is not done in VGE

**Comment by mewmew** [comment]
> Thanks for encouraging feedback :).
>
> Sorry, I don't have discord at the moment. However, I could try to install one after Easter holidays.

Happy to! Great, I'll send you my Discord username per e-mail. You can add me after the Easter holidays :)

**Comment by mewmew** [comment]
>  It is bit sad that Golang don't support similar DLL (so) loading in Linux that works in Windows. In most other areas Golang cross OS support is excellent.

I have a few tricks up my sleeve we can try after we manage to get it running on Linux. One approach could be to build `libvge` as an object archive, and then statically link the code into the Go binary. I've used this for other projects to avoid having to rely on a shared object/DLL file.

**Comment by mewmew** [comment]
I tried building https://github.com/SaschaWillems/Vulkan and the examples run successfully.

`./bin/triangle` (source [examples/triangle/triangle.cpp](https://github.com/SaschaWillems/Vulkan/blob/master/examples/triangle/triangle.cpp)):

![screenshot_2021-03-24_20:56:59](https://user-images.githubusercontent.com/1414531/112375193-49e2d780-8ce3-11eb-8378-7fab746f05fa.png)

**Comment by lakal3** [comment]
Could you try to debug PrepareSwapchain function in desktop.cpp file (around line 162) with GBD or LLDB debugger.

It will call createSwapchain function that should create swapchain images but for some reason we end up with zero images. Where VGE:s createSwapchains works differently from code in triangle.cpp.

I try to get/buy laptop with Linux and Intel integrated GPU but that might take some time. Without actual hardware it is very difficult to guess where something goes wrong. This code works with NVidia's native Linux drivers.

**Comment by mewmew** [comment]
Hi @lakal3,

Thanks for guiding the investigation into this issue!

I tried to do some initial debugging. I'm still quite new to Vulkan so, I tried to include as much relevant debug information as I could below, please let me know what aspects to continue to debug. It seems at the end of the function, the call to `getSwapchainImagesKHR` return `imCount` of 0. Do you know why this may be the case? The other parts look good, e..g `supported = 1` on line 280. The correct window dimensions seem to be used, etc.

I'd be happy to continue debugging, just need some guidance on where best to look as my Vulkan knowledge is still very novice.

Cheers,
Robin

# Info from `vge::Window::PrepareSwapchain`

```
pwndbg> p _swapChain
$3 = {
  static objectType = vk::ObjectType::eSwapchainKHR,
  static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSwapchainKHR,
  m_swapchainKHR = 0x7fffbc01a8d0
}
```

```
pwndbg> p *imageDesc
$6 = {
  Width = 1024,
  Height = 768,
  Depth = 1,
  Format = vk::Format::eB8G8R8A8Srgb,
  Layers = 1,
  MipLevels = 1
}
```

# Info from `vge::Window::createSwapchain`

Contents of `_crInfo` at start of function.
```
pwndbg> p _crInfo
$1 = {
  static allowDuplicate = false,
  static structureType = vk::StructureType::eSwapchainCreateInfoKHR,
  sType = vk::StructureType::eSwapchainCreateInfoKHR,
  pNext = 0x0,
  flags = {
    m_mask = 0
  },
  surface = {
    static objectType = vk::ObjectType::eSurfaceKHR,
    static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSurfaceKHR,
    m_surfaceKHR = 0x0
  },
  minImageCount = 0,
  imageFormat = vk::Format::eUndefined,
  imageColorSpace = vk::ColorSpaceKHR::eSrgbNonlinear,
  imageExtent = {
    width = 0,
    height = 0
  },
  imageArrayLayers = 0,
  imageUsage = {
    m_mask = 0
  },
  imageSharingMode = vk::SharingMode::eExclusive,
  queueFamilyIndexCount = 0,
  pQueueFamilyIndices = 0x0,
  preTransform = vk::SurfaceTransformFlagBitsKHR::eIdentity,
  compositeAlpha = vk::CompositeAlphaFlagBitsKHR::eOpaque,
  presentMode = vk::PresentModeKHR::eImmediate,
  clipped = 0,
  oldSwapchain = {
    static objectType = vk::ObjectType::eSwapchainKHR,
    static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSwapchainKHR,
    m_swapchainKHR = 0x0
  }
}
```

```c
 ► 280 		if (!supported) {
```

```
pwndbg> p supported
$4 = 1
```

```c
   287 		dev->get_pd().getSurfaceFormatsKHR(_surface, &sfCount, formats.data(), dev->get_dispatch());
 ► 288 		if (formats[0].format == vk::Format::eUndefined) {
```

```
pwndbg> p sfCount
$5 = 2
```

```c
 ► 312 	_crInfo.imageExtent.height = static_cast<uint32_t>(sfCap.currentExtent.height);
```

```
pwndbg> p sfCap
$7 = {
  minImageCount = 3,
  maxImageCount = 0,
  currentExtent = {
    width = 1024,
    height = 768
  },
  minImageExtent = {
    width = 1024,
    height = 768
  },
  maxImageExtent = {
    width = 1024,
    height = 768
  },
  maxImageArrayLayers = 1,
  supportedTransforms = {
    m_mask = 1
  },
  currentTransform = vk::SurfaceTransformFlagBitsKHR::eIdentity,
  supportedCompositeAlpha = {
    m_mask = 9
  },
  supportedUsageFlags = {
    m_mask = 31
  }
}
```

```c
   317 		_swapChain = dev->get_device().createSwapchainKHR(_crInfo, allocator, dev->get_dispatch());
```

```
pwndbg> p _swapChain
$9 = {
  static objectType = vk::ObjectType::eSwapchainKHR,
  static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSwapchainKHR,
  m_swapchainKHR = 0x8b38f0
}
```

Contents of `_crInfo` at end of function.
```
pwndbg> p _crInfo
$11 = {
  static allowDuplicate = false,
  static structureType = vk::StructureType::eSwapchainCreateInfoKHR,
  sType = vk::StructureType::eSwapchainCreateInfoKHR,
  pNext = 0x0,
  flags = {
    m_mask = 0
  },
  surface = {
    static objectType = vk::ObjectType::eSurfaceKHR,
    static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSurfaceKHR,
    m_surfaceKHR = 0x999df0
  },
  minImageCount = 0,
  imageFormat = vk::Format::eB8G8R8A8Srgb,
  imageColorSpace = vk::ColorSpaceKHR::eSrgbNonlinear,
  imageExtent = {
    width = 1024,
    height = 768
  },
  imageArrayLayers = 1,
  imageUsage = {
    m_mask = 17
  },
  imageSharingMode = vk::SharingMode::eExclusive,
  queueFamilyIndexCount = 0,
  pQueueFamilyIndices = 0x0,
  preTransform = vk::SurfaceTransformFlagBitsKHR::eIdentity,
  compositeAlpha = vk::CompositeAlphaFlagBitsKHR::eOpaque,
  presentMode = vk::PresentModeKHR::eFifo,
  clipped = 1,
  oldSwapchain = {
    static objectType = vk::ObjectType::eSwapchainKHR,
    static debugReportObjectType = vk::DebugReportObjectTypeEXT::eSwapchainKHR,
    m_swapchainKHR = 0x0
  }
}
```

```c
   319 	uint32_t imCount = 0;
   320 	dev->get_device().getSwapchainImagesKHR(_swapChain, &imCount, nullptr, dev->get_dispatch());
```

```
pwndbg> p imCount
$10 = 0
```

**Comment by mewmew** [comment]
```
pwndbg> p sfCap
$7 = {
  minImageCount = 3,
  maxImageCount = 0,
```

The `sfCap.maxImageCount` of 0 looks suspicious. This would also set `_crInfo.minImageCount` to 0 at line 295 of `desktop.cpp`:

```c
		if (_crInfo.minImageCount > sfCap.maxImageCount) {
			_crInfo.minImageCount = sfCap.maxImageCount;
		}
```

What values do you get for minImageCount and maxImageCount on your system @lakal3?

**Comment by lakal3** [comment]
OK. It is stated in vulkan spec:

maxImageCount is the maximum number of images the specified device supports for a swapchain created for the surface, and will be either 0, or greater than or equal to minImageCount. A value of 0 means that there is no limit on the number of images, though there may be limits related to the total amount of memory used by presentable images.

Well like with any user manuals, you only read specification when nothing else helps.

Is just pushed a "blind" fix that I cannot verify. Can you check it that works?

**Comment by mewmew** [comment]
Fixed in 37ce3b96aebf1f9791a0fced631dd8e3ed321b1b. Thanks for helping me troubleshoot this issue and find a clean solution.
